tcf_website/management/commands/load_grades.py

```diff
@@ -0,0 +1,398 @@
+                'D-',
+                'F'])
+
+    def load_semester_file(self, file):
+        year, semester = file.split('.')[0].split('_')
+        year = int(year)
+        season = semester.upper()
+
+        print(year, season)
+
+        df = self.clean(pd.read_excel(os.path.join(self.data_dir, file)))
+
+        print(f"{df.size} sections")
+
+        for index, row in tqdm(df.iterrows(), total=df.shape[0]):
+            # print(str(row).encode('ascii', 'ignore').decode('ascii'))
+            self.load_row_into_dict(row)
+            # break
+
+        self.load_dict_into_models()
+
+    def load_row_into_dict(self, row):
+        try:
+            # parsing fields of excel
+
+            first_name = row['Instructor First Name']
+            middle_name = row['Instructor Middle Name']
+            last_name = row['Instructor Last Name']
+            email = row['Instructor Email']
+            subdepartment = row['Subject']
+            number = re.sub('[^0-9]', '', str(row['Course Number']))
+            section_number = row['Section Number']  # not used
+            title = row['Title']
+            gpa = float(row['Course GPA'])
+            a_plus = int(row['A+'])
+            a = int(row['A'])
+            a_minus = int(row['A-'])
+            b_plus = int(row['B+'])
+            b = int(row['B'])
+            b_minus = int(row['B-'])
+            c_plus = int(row['C+'])
+            c = int(row['C'])
+            c_minus = int(row['C-'])
+            d_plus = int(row['D+'])
+            d = int(row['D'])
+            d_minus = int(row['D-'])
+            f = int(row['F'])
+            ot = int(row['OT'])
+            drop = int(row['DR'])
+            withdraw = int(row['W'])
+            # credit
+            # general_credit
+            # no_credit
+            total_enrolled = int(row['Total'])
+
+            # identifiers are tuple keys to dictionaries
+            course_identifier = (subdepartment, number, title)
+            course_instructor_identifier = (
+                subdepartment, number, first_name, middle_name, last_name, email)
+
+            # value of dictionaries (incremented onto value if key already
+            # exists)
+            this_semesters_grades = [
+                a_plus,
+                a,
+                a_minus,
+                b_plus,
+                b,
+                b_minus,
+                c_plus,
+                c,
+                c_minus,
+                d_plus,
+                d,
+                d_minus,
+                f,
+                ot,
+                drop,
+                withdraw]
+
+            # load this semester into course dictionary
+            if course_identifier not in course_grades:
+                course_grades[course_identifier] = this_semesters_grades
+            else:
+                for i in range(len(course_grades[course_identifier])):
+                    course_grades[course_identifier][i] += this_semesters_grades[i]
```

Comment:
General concern is that this overlaps data for Special Topics (e.g. CS 1501 or whatever) courses...but I can't think of a meaningful way to fix this.

Reply:
I don't think anyone can fix this because this isn't something SIS considered in the beginning. Sherriff and McBurney both teach CS 3240 this semester, and their classes are considered a single class. That is, students are in different sections but listen to the same lectures and are graded on the same scale. On the other hand, Shen and Magee both teach CS 4740 this semester, but the only thing shared between the two classes are the name of the class. They are taught different materials, have different assignments, and graded on different scales. Ideally, we need another identifier for each section, and which should allow us to group different sections based on that, but we don't have that data.